extensions/amp-instagram/1.0/component.js

```diff
@@ -15,34 +15,73 @@
  */
 
 import * as Preact from '../../../src/preact';
-import {ContainWrapper} from '../../../src/preact/component';
+import {ContainWrapper, useValueRef} from '../../../src/preact/component';
+import {Loading} from '../../../src/loading';
+import {ReadyState} from '../../../src/ready-state';
 import {dict} from '../../../src/utils/object';
+import {forwardRef} from '../../../src/preact/compat';
 import {getData} from '../../../src/event-helper';
 import {parseJson} from '../../../src/json';
-import {useLayoutEffect, useRef, useState} from '../../../src/preact';
-import {useLoad} from '../../../src/preact/context';
+import {
+  useImperativeHandle,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useLoading} from '../../../src/preact/context';
 
 const NO_HEIGHT_STYLE = dict();
 
 /**
- * @param {!InstagramPropsDef} props
+ * @param {!InstagramDef.Props} props
+ * @param {{current: (!InstagramDef.Api|null)}} ref
  * @return {PreactDef.Renderable}
+ * @template T
  */
-export function Instagram({
-  shortcode,
-  captioned,
-  title,
-  requestResize,
-  loading,
-  onLoad,
-  ...rest
-}) {
-  const load = useLoad(loading, /* unlayoutOnPause */ true);
+export function InstagramWithRef(
+  {
+    shortcode,
+    captioned,
+    title,
+    requestResize,
+    loading: loadingProp,
+    onReadyState,
+    ...rest
+  },
+  ref
+) {
+  const loading = useLoading(loadingProp, /* unlayoutOnPause */ true);
+  const load = loading !== Loading.UNLOAD;
+  const [loaded, setLoaded] = useState(false);
 
   const iframeRef = useRef(null);
   const [heightStyle, setHeightStyle] = useState(NO_HEIGHT_STYLE);
   const [opacity, setOpacity] = useState(0);
 
+  // Component API: InstagramDef.Api.
+  useImperativeHandle(
+    ref,
+    () => ({
+      // Standard Bento
+      get readyState() {
+        return loaded ? ReadyState.COMPLETE : ReadyState.LOADING;
+      },
+    }),
+    [loaded]
+  );
+
+  // Reset readyState to "laoding" when an iframe is unloaded.
+  const onReadyStateRef = useValueRef(onReadyState);
+  useLayoutEffect(() => {
+    if (!load) {
+      setLoaded(false);
+      const onReadyState = onReadyStateRef.current;
+      if (onReadyState) {
+        onReadyState(ReadyState.LOADING);
+      }
+    }
```

Comment:
Right. This is why `useLayoutEffect` was important. Basically the risk here is a race condition. We may unload an iframe and schedule the effect for it. But then it could also get reloaded and it'd send its "load" event. The race condition could be between `useEffect(() => setReadyState("loading"))` and `iframe.onload = () => setReadtState("complete")`. The `useLayoutEffect` prevents the race condition.

On the other hand, we can't have an affect that modify the state that it depends on - that might cause the infinite loop.

However, as a compromise I refactored this a bit and moved all `loaded` and event handling into `setLoaded` callback. PTAL and let me know if that's better.